qiskit_experiments/library/characterization/tphi.py

```diff
@@ -0,0 +1,119 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021, 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Tphi Experiment class.
+"""
+
+from typing import List, Optional, Union
+import numpy as np
+
+from qiskit import QiskitError
+from qiskit.providers import Backend
+from qiskit.test.mock import FakeBackend
+from qiskit_experiments.framework.composite.batch_experiment import BatchExperiment
+from qiskit_experiments.library.characterization import (
+    T1,
+    T2Ramsey,
+    TphiAnalysis,
+    T1Analysis,
+    T2RamseyAnalysis,
+)
+
+
+class Tphi(BatchExperiment):
+    r"""Tphi Experiment Class
+
+    # section: overview
+
+        :math:`\Gamma_\varphi` is defined as the rate of pure dephasing
+        or depolarization in the :math:`x - y` plane.
+        We compute :math:`\Gamma_\varphi` by computing :math:`\Gamma_2*`, the transverse relaxation rate,
+        and subtracting :math:`\Gamma_1`, the longitudinal relaxation rate. The pure dephasing time
+        is defined by :math:`T_\varphi = 1/\Gamma_\varphi`. Or more precisely,
+
+        :math:`1/T_\varphi = 1/T_{2*} - 1/2T_1`.
+
+        For more details, see :py:class:`T1` and :py:class:`T2Ramsey`
+
+    # section: analysis_ref
+        :py:class:`TphiAnalysis`
+
+    # section: reference
+        .. ref_arxiv:: 1 1904.06560v5
+
+    # section: tutorial
+        :doc:`/tutorials/tphi_characterization`
+    """
+
+    def set_experiment_options(self, **fields):
+        """Set the experiment options.
+        Args:
+            fields: The fields defining the options
+
+        Raises:
+             QiskitError : Error for invalid input option.
+        """
+        # propagate options to the sub-experiments.
+        for key in fields:
+            if key == "delays_t1":
+                self.component_experiment(0).set_experiment_options(delays=fields["delays_t1"])
```

Comment:
There's a subtle point here, and I think what you're doing is the most appropriate.

The experiment options are exposed to the user by the property `BaseExperiment.experiment_options`. I understand that you wish to mask the internal implementation of a batch experiment consisting of sub-experiments. But a user who'll try `tphi_exp.experiment_options.delays_t1` will get an error, because Tphi doesn't really have an option named `delays_t1`.

You can try to solve this by maintaining options `delays_t1` and `delays_t2` in Tphi, in addition to the delay options of the sub-experiments. But then what will you do if the user does `tphi_exp.experiment_options.delays_t1 = ...`. You'll face again the issue of having to sync with the delays of the sub-experiment.

You'll keep facing similar issues if you try to override `experiment_options`. Possibly the only way to obtain full masking is by overriding `Options`, which is exaggerated. Hence the bottom line (which I already wrote in the upper line) that the current code is the most reasonable one, in my opinion.

Reply:
If this were C++, I could have one of the delays_t2 (in Tphi and in T2Ramsey) fields point to the other. But as far as I know, there is no pointer concept in Python, right?